modules/free/src/main/scala/doobie/WeakAsync.scala

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2013-2020 Rob Norris and Contributors
+// This software is licensed under the MIT License (MIT).
+// For more information see LICENSE or https://opensource.org/licenses/MIT
+
+package doobie
+
+import cats.~>
+import cats.implicits._
+import cats.effect.kernel.{ Async, Poll, Resource, Sync }
+import cats.effect.implicits._
+import cats.effect.std.Dispatcher
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+trait WeakAsync[F[_]] extends Sync[F] {
+  def fromFuture[A](fut: F[Future[A]]): F[A]
+}
+
+object WeakAsync {
```

Comment:
Would it be worth to put a comment in the code explaining why `WeakAsync` is necessary? From what I can see it's `Sync` but with `fromFuture` and nothing else. Why not just use `Async` directly?

Reply:
Absolutely. The short explanation is that `Async` cannot be implemented for `ConnectionIO` because it implies concurrency. You cannot concurrently access a `Connection` or a `ResultSet` from different fibers / threads.

One the other hand `Sync` is not sufficient, because it would preclude streaming data produced with `IO` effects into the database. In order to do that we need a `FunctionK` from `IO` to `ConnectionIO`. Cats Effect 3 `Dispatcher` can be uses do that, but only if `ConnectionIO` can absorb a value that is produced asynchronously. Hence `WeakAsync.fromFuture`.

See also https://github.com/rkrzewski/doobie/blob/cats-effect-3/modules/free/src/main/scala/doobie/WeakAsync.scala#L45